### PR TITLE
[JVM] Necromance a cheaper (de)serialization

### PR DESCRIPTION
**Pull request by Kaiepi**
This started as means of fixing https://github.com/rakudo/rakudo/issues/4952 for the next stretch of https://github.com/Raku/nqp/pull/776, but is kind of its own rabbit hole. The `hs_err` log produced by the failure in `CORE.d.setting` points to this line as the source of the error from within nqp:

https://github.com/Raku/nqp/blob/4f4d4ef1f6776a0a169441f18060ad6863ff5110/src/vm/jvm/runtime/org/raku/nqp/runtime/LibraryLoader.java#L71

While looking *around* here, I noticed we are very lackadaisical with how we handle serialized bytecode within memory. Namely, when `LibraryLoad.load` is invoked, we can wind up copying entire files in memory *multiple times* over, and their decompressed classfile and bytecode files persist in memory *as long as their classes do*. I haven't quite managed to even provoke a change in the `CORE.d.setting` error produced, but I have managed to convince `CORE.c.setting` to build in 90% of the time it does on master by reducing the copying to just one instance so far. This can be taken further to eliminate copying of files in bulk entirely, but will take time.

Passes `make test` with OpenJDK 19-ea, I guess.

### CODE REVIEW

**Comment by Kaiepi** [comment]
Turns out `-XX:+UseTransparentHugePages` is a detriment! `CORE.c.setting` can now build 23% faster than on master.

**Comment by Kaiepi** [comment]
There's more that can be done in this vein in the `SerializationReader` whereabouts, but I'd rather not mess with both the library loader and serialization in general at the same time. CI seems to enjoy these changes as-is.

**Comment by usev6** [comment]
This looks/sounds very promising.

I don't feel qualified to give it a real review, but I'll at least try to test the branch. (But that will take some time.)

**Comment by patrickbkr** [comment]
Given this unblocks Kaiepi, what hinders merging this as is? It doesn't seem to touch MoarVM things, so can't break thinks too badly.

File: src/vm/jvm/runtime/org/raku/nqp/runtime/GlobalContext.java [hunk 1]
```diff
@@ -4,8 +4,9 @@
 import java.io.PrintStream;
 import java.io.UnsupportedEncodingException;
 import java.lang.ref.WeakReference;
+import java.util.concurrent.CopyOnWriteArraySet;
```
Comment (usev6) [inline src/vm/jvm/runtime/org/raku/nqp/runtime/GlobalContext.java:7]: If I'm not mistaken, this import isn't needed after https://github.com/Kaiepi/nqp/commit/5d026b44ba

File: src/vm/jvm/runtime/org/raku/nqp/runtime/GlobalContext.java [hunk 2]
```diff
@@ -4,8 +4,9 @@
...
 import java.util.HashMap;
-import java.util.HashSet;
+import java.util.Set;
```
Comment (usev6) [inline src/vm/jvm/runtime/org/raku/nqp/runtime/GlobalContext.java:9]: If I'm not mistaken, this import isn't needed after https://github.com/Kaiepi/nqp/commit/5d026b44ba

File: tools/build/install-jvm-runner.pl.in [hunk 3]
```diff
@@ -21,15 +21,15 @@ if ($^O eq 'MSWin32') {
     open my $fh, ">", $install_to
         or die "Could not open $install_to: $!";
     print $fh '@java -Xss1m -Xmx1324m -Xbootclasspath/a:' . $jar_dir . '\\nqp-runtime.jar;' .
-              "$jar_dir\\@asmfile@;$jar_dir\\@jlinefile@;$jar_dir\\@jnafile@;$jar_dir\\@jnaplatformfile@;$lib_dir\\nqp.jar -cp $lib_dir nqp %*\n";
+              "$jar_dir\\@asmfile@;$jar_dir\\@jlinefile@;$jar_dir\\@jnafile@;$jar_dir\\@jnaplatformfile@;$jar_dir\\@xzfile@;$lib_dir\\nqp.jar -cp $lib_dir nqp %*\n";
```
[the remark this answers is not in the document]
Comment (usev6) [reply tools/build/install-jvm-runner.pl.in:24]: But to be honest, I didn't figure out yet, where this is used.

File: src/vm/jvm/runtime/org/raku/nqp/runtime/LibraryLoader.java [hunk 4]
```diff
@@ -40,138 +52,423 @@ public void load(ThreadContext tc, String origFilename) {
...
+        try (JarInputStream jis = new JarInputStream(new ByteBufferedInputStream(buffer))) {
+            JarEntry je;
+            while ((je = jis.getNextJarEntry()) != null) {
+                String jf = je.getName();
+                if (jf.endsWith(".class") && classfile == null) {
+                    name = je.getComment();
+                    classfile = readToHeapBuffer(jis);
+                }
+                else if (jf.endsWith(".serialized.lz4") && serial == null)
+                    serial = readToHeapBuffer(jis);
```
Comment (usev6) [inline src/vm/jvm/runtime/org/raku/nqp/runtime/LibraryLoader.java:138]: Likely I'm missing something, but shouldn't this be ```readToHeapBufferLz4(jis)```?

**Review by usev6 — COMMENTED** [review]
As I noted earlier, I don't feel qualified for a real review. But I tried my best and raised a few questions. @Kaiepi maybe you could have a look at those?

Apart from those comments this still looks very promising to me. I didn't come around to give it a real test, but will do that now.

File: src/vm/jvm/runtime/org/raku/nqp/runtime/LibraryLoader.java (hunk 4 above)
Comment (Kaiepi) [reply src/vm/jvm/runtime/org/raku/nqp/runtime/LibraryLoader.java:138]: It's been a while, but I think the burden of decompressing this is left on whatever depends on `MemoryClassLoader`'s `getResourceAsStream`. This is done to spare memory in its copying, but maybe there's a way to not do that to begin with.

Comment (usev6) [reply src/vm/jvm/runtime/org/raku/nqp/runtime/LibraryLoader.java:138]: I see, so it wasn't an oversight. Thanks for explaining :)

**Comment by usev6** [comment]
A spectest looked good as well.

**Comment by usev6** [comment]
> Given this unblocks Kaiepi, what hinders merging this as is? It doesn't seem to touch MoarVM things, so can't break thinks too badly.

From me it's a +1 for merging.

If no-one else chimes in I'll hit the merge button in a couple of days.

**Pull request merged**